### PR TITLE
Enable addon registration

### PR DESCRIPTION
**Pull request by elgnay**
The related task: https://github.com/open-cluster-management/backlog/issues/11279

The enhancement proposal: https://github.com/open-cluster-management/enhancements/pull/13

Signed-off-by: Yang Le <yangle@redhat.com>

### CODE REVIEW

File: pkg/controller/managedclusteraddon/managed_cluster_addon_registration.go [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+		config,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// UpdateRegisrationConfig adds registration annotations to ManagedClusterAddOn
```
Comment (qiujian16) [inline pkg/controller/managedclusteraddon/managed_cluster_addon_registration.go:262]: add a todo to reflect new apis.

File: pkg/controller/managedclusteraddon/managed_cluster_addon_registration.go [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+	err = newApplier.CreateOrUpdateInPath(
+		"resources/hub/common",
+		nil,
+		false,
+		config,
+	)
+	if err != nil {
+		return err
+	}
+	// delete old role & rolebindings created in previous releases
```
Comment (qiujian16) [inline pkg/controller/managedclusteraddon/managed_cluster_addon_registration.go:117]: Im not sure we really need to do this.  I would remove deletion from this PR. The concern is mainly on version skew.

File: pkg/controller/csr/csr_controller.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+		Reason:  "AutoApprovedByHubCSRController",
+		Message: "Auto approving addon certificate signing request.",
+		Status:  corev1.ConditionTrue,
+	})
+	_, err := r.csrClient.UpdateApproval(context.TODO(), csr.Name, csr, metav1.UpdateOptions{})
+	return err
+}
+
+func isValidAddonCSR(csr *certificatesv1.CertificateSigningRequest, managedClusterAddonName, clusterName string) bool {
+	block, _ := pem.Decode(csr.Spec.Request)
```
Comment (qiujian16) [inline pkg/controller/csr/csr_controller.go:183]: We should also check the requester.

File: pkg/controller/csr/csr_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,269 @@
+// (c) Copyright IBM Corporation 2019, 2020. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (elgnay) [reply pkg/controller/csr/csr_controller_test.go]: Copyright updated.

File: pkg/controller/managedclusteraddon/managed_cluster_addon_registration.go (hunk 1 above)
Comment (elgnay) [reply pkg/controller/managedclusteraddon/managed_cluster_addon_registration.go:234]: If possible I'd like to update this pr to use the new apis.

File: pkg/controller/csr/csr_controller.go (hunk 3 above)
Comment (elgnay) [reply pkg/controller/csr/csr_controller.go:178]: Requester is checked by predicate.

**Comment by hanqiuzh** [comment]
/approve

leave lgtm to @leena-jawale

**Pull request merged**